File: twilio/packages/dialplans/metric_dialer_status/sns-client.js
// Functions to publish metric events to SNS.

const { SNSClient, PublishCommand } = require("@aws-sdk/client-sns");

const futelUtil = require('./futel-util');
const metricHostBase = 'twilio-sip-server';

// Return the appropriate metric event hostname for our environment.
function getMetricHostname(context) {
    return metricHostBase + '-' + futelUtil.getEnvironment(context);
}

function eventToMessage(context, event) {
    let dateString = new Date().toISOString();
    // PJSIP is an implementation detail from Asterisk that the usage grapher might require?
    //let channelString = "PSJIP/" + event.Channel;
    event = {
        Event: 'UserEvent',
        Channel: event.Channel,
        UserEvent: event.UserEvent
    };
    message = {
        timestamp: dateString,
        hostname: getMetricHostname(context),
        event: event};
    return JSON.stringify(message);
}

// Return a config object populated from environment.
function getConfig() {
    return {
        region: 'us-west-2',
        credentials: {
            accessKeyId: process.env.AWS_ACCESS_KEY_ID,
            secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY}};
}

// Return a publish parameters object populated from environment and message.
function getPublishParams(message) {
    return {
        Message: message,
        TopicArn: process.env.AWS_TOPIC_ARN
    };
}

// Return a Promise with the response from publishing the event to SNS.
function publish(context, event, topic_arn) {
    message = eventToMessage(context, event);
    const client = new SNSClient(getConfig());
    const command = new PublishCommand(getPublishParams(message));
    return client.send(command);
}

exports.publish = publish;

File: twilio/packages/dialplans/metric_dialer_status/sns_client.py
import boto3
import datetime
import json
import os

import util


metric_host_base = 'do-dialplans';

def getenv():
    return {
        'AWS_ACCESS_KEY_ID': os.environ['AWS_ACCESS_KEY_ID'],
        'AWS_SECRET_ACCESS_KEY': os.environ['AWS_SECRET_ACCESS_KEY'],
        'AWS_TOPIC_ARN': os.environ['AWS_TOPIC_ARN']}

# Return the appropriate metric event hostname for our environment.
def get_metric_hostname():
    return metric_host_base + '-' + util.get_instance()

def event_to_message(event, hostname):
    date_string = datetime.datetime.now().isoformat()
    event = {
        'Event': 'UserEvent',
        'Channel': event['channel'],
        'UserEvent': event['user_event']}
    message = {
        'timestamp': date_string,
        'hostname': hostname,
        'event': event}
    return message

def publish(event, env):
    hostname = get_metric_hostname()
    message = event_to_message(event, hostname)
    client = boto3.client(
        'sns',
        aws_access_key_id=env['AWS_ACCESS_KEY_ID'],
        aws_secret_access_key=env['AWS_SECRET_ACCESS_KEY'],
        region_name= 'us-west-2')
    return client.publish(
        TargetArn=env['AWS_TOPIC_ARN'],
        Message=json.dumps({'default': json.dumps(message)}),
        MessageStructure='json')
